src/buffer/assert.rs

```diff
@@ -1,57 +1,36 @@
+#[deprecated = "use actual.assert_eq(&expected)"]
 #[macro_export]
 macro_rules! assert_buffer_eq {
     ($actual_expr:expr, $expected_expr:expr) => {
-        match (&$actual_expr, &$expected_expr) {
-            (actual, expected) => {
-                if actual.area != expected.area {
-                    panic!(
-                        indoc::indoc!(
-                            "
-                            buffer areas not equal
-                            expected:  {:?}
-                            actual:    {:?}"
-                        ),
-                        expected, actual
-                    );
-                }
-                let diff = expected.diff(&actual);
-                if !diff.is_empty() {
-                    let nice_diff = diff
-                        .iter()
-                        .enumerate()
-                        .map(|(i, (x, y, cell))| {
-                            let expected_cell = expected.get(*x, *y);
-                            indoc::formatdoc! {"
-                                {i}: at ({x}, {y})
-                                  expected: {expected_cell:?}
-                                  actual:   {cell:?}
-                            "}
-                        })
-                        .collect::<Vec<String>>()
-                        .join("\n");
-                    panic!(
-                        indoc::indoc!(
-                            "
-                            buffer contents not equal
-                            expected: {:?}
-                            actual: {:?}
-                            diff:
-                            {}"
-                        ),
-                        expected, actual, nice_diff
-                    );
-                }
-                // shouldn't get here, but this guards against future behavior
-                // that changes equality but not area or content
-                assert_eq!(actual, expected, "buffers not equal");
-            }
-        }
+        let (actual, expected) = (&$actual_expr, &$expected_expr);
+        assert!(
+            actual.area == expected.area,
+            "buffer areas not equal\nexpected: {expected:?}\nactual:   {actual:?}",
+        );
+        let nice_diff = expected
+            .diff(actual)
+            .into_iter()
+            .enumerate()
+            .map(|(i, (x, y, cell))| {
+                let expected_cell = expected.get(x, y);
+                format!("{i}: at ({x}, {y})\n  expected: {expected_cell:?}\n  actual:   {cell:?}")
+            })
+            .collect::<Vec<String>>()
+            .join("\n");
+        assert!(
+            nice_diff.is_empty(),
+            "buffer contents not equal\nexpected: {expected:?}\nactual:   {actual:?}\ndiff:\n{nice_diff}",
+        );
+        // shouldn't get here, but this guards against future behavior
+        // that changes equality but not area or content
+        assert_eq!(actual, expected, "buffers not equal");
```

Comment:
Should the macro just use the assert_eq? I started with simplifying this but now its completely moved to the assert_eq function anyway so prevent code duplication until this deprecated macro is removed?

```suggestion
        actual.assert_eq!(expected);
```

Reply:
The entire rationale for this macro existing is built around two ideas:
1. When a test fails, the information reported should be enough to understand exactly how the buffers differ
2. When writing new tests, the output of the assertion should make it easy to write an intentionally failing assertion (`assert_buffer_eq!(actual, Buffer::empty(actual.area))`) and directly copy the result into the test once confirmed to be correct (with the exact number of spaces, newlines and quotes to make this seemless).

So long as both of those items are true, how the macro / function call works is mostly irrelevant.

It's possible that the first idea could be met by providing a better `Debug` representation that makes each line be displayed in a way that makes the style information easy to read and also fully copy-paste friendly. E.g. something like the following:
```rust
[
    Line::raw("A"),
    Line::styled("B", Color::Blue),
    Line::from([Span::raw("C"), Span::styled("D", Color::Red]),
    Line::styled("E", (Color::Black, Color::White, Modifier::BOLD | Modfier::Italic)),
].into()
```

If the debug output was that good, then we could just use `assert_eq!` throughout our tests instead of having a specific macro / function call.

This makes me think that perhaps it's worth making this function pub(crate) for now and having a stretch to get to the good debug output like that.

Reply:
> 2\. When writing new tests, the output of the assertion should make it easy to write an intentionally failing assertion (`assert_buffer_eq!(actual, Buffer::empty(actual.area))`) and directly copy the result into the test once confirmed to be correct (with the exact number of spaces, newlines and quotes to make this seemless).

This is done by the custom Debug implementation already. The assert_buffer_eq macro doesnt do anything special for that case.

> If the debug output was that good, then we could just use `assert_eq!` throughout our tests instead of having a specific macro / function call.

I like that idea! Sadly `Buffer::diff` is used in the assertion to highlight the differences better. We can not achieve something like this with just the Debug implementation as it only knows one of the two buffers. Also, it would just notice the two buffers being different and then print the whole Debug output of the two buffers which isn't as easy to compare.

> How does the output look if you replace this with `actual.assert_eq`? If it's sane, then yep go for it.

The code is exactly the same, its duplicated. The output is also exactly the same. It's also the same compared to the main branch output.

Reply:
Settled on assert_eq and a good Debug impl.